README.md
# replicated-lint

[![CircleCI](https://circleci.com/gh/replicatedhq/replicated-lint/tree/master.svg?style=svg&circle-token=9ae2573df7075cff352d329eb7f88d52037872b5)](https://circleci.com/gh/replicatedhq/replicated-lint/tree/main)
[![Code Climate](https://codeclimate.com/github/replicatedhq/replicated-lint/badges/gpa.svg)](https://codeclimate.com/github/replicatedhq/replicated-lint)
[![Test Coverage](https://codeclimate.com/github/replicatedhq/replicated-lint/badges/coverage.svg)](https://codeclimate.com/github/replicatedhq/replicated-lint)

YAML linting tools for Replicated applications.

## Usage

Install the CLI executable with

```sh
npm install -g replicated-lint
```

Lint with `replicated-lint validate`

```sh
replicated-lint validate -f my-app.yml
```

or pipe from stdin:

```sh
cat my-app.yml | replicated-lint validate -f -
```

Results that have issues will look something like:

```typescript
{ type: 'info',
  rule: 'prop-configitem-testproc-run-on-save',
  message: 'If a config item\'s test_proc.run_on_save is not set to \'true\', test_proc\'s will not be checked automatically. Consider setting your test_proc\'s run_on_save to automatically validate inputs',
  positions:
   [ { path: 'config.1.items.2.test_proc',
       start: { position: 8130, line: 325, column: 4 },
       end: { position: 8322, line: 331, column: 0 } },
     { path: 'config.3.test_proc',
       start: { position: 8692, line: 346, column: 2 },
       end: { position: 9141, line: 365, column: 2 } } ],
  links: [ 'https://www.replicated.com/docs/packaging-an-application/test-procs/' ] }

# prop-configitem-testproc-run-on-save continued from line 321
322  
323    - name: phone_number
324      type: text
325      test_proc:
326        display_name: Is this a Phone Number?
327        command: regex_match
328        args:
329        - "([0-9]{3})[-]([0-9]{3})[-]([0-9]{4})$"
330        - "That doesn't seem to be a phone number!"
331  - name: auth
332    title: Authentication
333    description: Where will user accounts be provisioned
334    items:


```

### Extending the CLI with custom rules

`replicated-lint` rules can be expressed as JSON, so it is easy to add your own custom rules.

If you have a custom rule set in `no-latest.json`, you can pass it to `replicated-lint` using

```sh
cat my-app.yml | replicated-lint validate -f - --extraRules no-latest.yaml
```

`--extraRules` can be specified multiple times. An example YAML rule set might look something like

```yaml
---
- name: custom-no-latest
  type: error
  message: "Don't use `latest` for container versions."
  test:
    AnyOf:
      path: components
      pred:
        AnyOf: containers
        pred:
          Eq:
            path: version
            value: latest
```

`replicated-lint validate` supports the following options:

```
Options:
  --version         Show version number                                [boolean]
  --help            Show help                                          [boolean]
  --infile, -f      Input file to validate. Use "-" for stdin
                                                         [string] [default: "-"]
  --threshold, -t   Threshold of of issues to report
                  [string] [choices: "info", "warn", "error"] [default: "error"]
  --extraRules, -e  Path to file containing JSON definitions for additional yaml
                    rules. Can be specified multiple times.[array] [default: []]
  --reporter, -r    Output Format to use
                     [string] [choices: "console", "junit"] [default: "console"]
  --outputDir, -o   junit reporter only -- path to directory to output junit xml
                    reports                   [string] [default: "test-results"]
```

## Developing


### Installing Dependencies

We compile to ES5 and test on nodejs 7.8.0, but earlier versions of node should work as well.

```sh
npm install -g yarn
yarn
```

### Running the tests

To run the tests once

```sh
yarn test
```

To watch files and re-run tests on changes, use the `tdd` script

```sh
./tdd
```

### Before committing

Run

```sh
make project-import PROJECT=<project-name>
```

Where `<project-name>` is the project being updated and should be one of
 - replicated-supportbundle
 - replicated-entitlements
 - replicated-rbac
 - replicated-ship

 For more up to date list check the project directory listing.


### Regenerating the documentation

`replicated-lint docs gen` will print reference documentation stdout as markdown. `yarn docs` will write it to `docs/gen.md`.

To update the docs in help center, grab everything in `replicated-lint/docs/gen.md` and paste it into `help-center/.../yaml.md`, replacing everything below `<!-- the rest of this document is autogenerated-->`.

## Library Usage

`replicated-lint` is designed with extensibility in mind, making it suitable for general purpose
YAML and JSON linting and policy definition

- Flexible, modular rule sets
- CLI
- Autogenerate unit tests and documentation from rule definitions

```sh
npm install --save replicated-lint
```

```typescript
import * as linter from "replicated-lint";

const yaml = `
---
replicated_api_version: 2.8.0
components:
  - name: ELK
    containers:
      - image: getelk/search`;


const ruleViolations = linter.defaultLint(yaml);
console.log(ruleViolations); // []

```

### Custom Rule Sets

`linter.rules.all` can be substituted or extended with custom rule
definitions. A rule's `test` field should return `{ matched: true }`
when the rule is triggered by invalid JSON.

```typescript
import * as linter from "replicated-lint";

const yaml = `
---
foo:
  bar: baz`;

const testSpec: any = {
  Or: {
    preds: [
      {Eq: { path: "foo.bar", value: "baz"}},
      {Eq: { path: "foo.bar", value: "boz"}},
    ],
  }
}

const rules: linter.YAMLRule[] = [
  {
    name: "foo-bar-neq-baz",
    message: "foo.bar can't be baz!",
    test: testSpec,
    type: "error",
  }
];

const ruleViolations = linter.lint(yaml, { rules });
console.log(ruleViolations);  /*
[{
        type: "error",
        positions: [{
          path: "foo.bar",
          start: {
            position: 12,
            line: 3,
            column: 2,
          },
          end: {
            position: 20,
            line: 3,
            column: 10,
          },
        }],
        message: "foo.bar can't be baz!",
}]
*/
```

Register new rules with `linter.enginer.register`. Rules should implement `JSONReadable<Predicate<any>>`, usually as a static method

```typescript
import * as linter from "replicated-lint";


// rule MyRule checks if root object has property "spam" equal to "eggs"
class MyRule implements linter.Predicate<any> {
  test(obj: any) {
    const matched = obj.spam !== "eggs"; // fail when spam != eggs
    const paths = ["spam"];              // if matched == true, optionally include a path where a rule was violated
    return { matched, paths };
  }

  public static fromJson(obj: any, registry: linter.engine.Registry) {
    return new MyRule();
  }
}

linter.engine.register(MyRule);

const testSpec: linter.Test = { MyRule: {}};

const rules: linter.YAMLRule[] = [
  {
    name: "spam-eq-eggs",
    message: "spam must be equal to eggs!",
    test: testSpec,
    type: "error",
  }
];

const yaml = `
---
spam: eggs`;

const ruleViolations = linter.lint(yaml, { rules });
console.log(ruleViolations); // []
```

Custom implementations of `engine.Registry` can also be passed as a third argument to `linter.lint`, otherwise a default
registry will be used.


CHANGELOG
0.1.0 / 2017-06-09
    * Initial release

0.3.0 / 2017-06-17
    * Implementation of ruleset, accept raw json and compile to rule predicates

0.3.1 / 2017-06-17
    * Fix bug in testproc rule that caused it to fire when `test_proc` was not present on the config item/group

0.3.2 / 2017-06-17
    * Fix bug TraverseSearcher and fix an issue where Monitor predicate not registered

0.4.0 / 2017-06-17
    * Update rule definition schema, rewrite rules to match
    * Update rule messaging descriptions and examples

0.5.0 / 2017-07-11
    * Update rule definition schema and compilation to use a typed object, rewrite rules to match

0.6.0 / 2017-08-11
    * Incorporate json schema for validation
    * (breaking) `linter.lint()` now takes a `LinterOpts` object instead of positional params
        * `linter.lint(yaml, linter.rules.all)` should change to `linter.lint(yaml, {rules: linter.rules.all})`
    * Add `linter.lintDefault()` which will lint with the recommended schema and rules
    * Tweaks to `replicated-lint` bin command in package.json

0.6.1 / 2017-08-14
    * Port config item `type` validation https://github.com/replicatedhq/libyaml/blob/master/validate.go#L376
    * Port config item `when` validation https://github.com/replicatedhq/libyaml/blob/master/validate.go#L402
    * Remove `received` field from linter results.

0.6.2 / 2017-08-14
    * Port custom monitor `target` validation https://github.com/replicatedhq/libyaml/blob/master/validate.go#L487

0.6.3 / 2017-08-14
    * Port host requirements docker version validation https://github.com/replicatedhq/libyaml/blob/master/validate.go#L659

0.6.4 / 2017-08-14
    * Validate `kubernetes.requirements.server_version` is a valid semver range
    * Validate `host_requirements.replicated_version` is a valid semver range

0.6.5 / 2017-08-14
    * Validate `admin_commands.component` and `admin_commands.container` match a valid component/container
    * Fix a bug in `monitors.cpuacct` and `monitors.memory` validation where the linter would fail with no components, even if both `cpuacct` and `memacct` are empty

0.6.6 / 2017-08-15
    * Allow `support.timeout` to be either a `number` or `string`
    * Add a path to schema error messaging

0.6.7 / 2017-08-18
    * Messaging tweak for `test_proc` rule
    * Schema support for inline component/container in `admin_commands` and `support`
    * Skip validating container references in non-native scheduler sources in `admin_commands` and `support`

0.6.8 / 2017-08-18
    * Validate that `components.*.containers.*.publish_events.*.subscriptions` all reference valid component/containers

0.6.9 / 2017-08-18
    * Validate that a component's cluster `strategy` is one of `autoscale` or `random`

0.6.10 / 2017-08-18
    * Ensure no conflicting volume options are specified

0.6.11 / 2017-08-18
    * Ensure host_path and container_path are absolute paths
    * Validate custom monitor colors

0.6.12 / 2017-08-21
    * Ensure `graphite.port` and `statsd.port` are valid tcp ports
    * Ensure `custom_metrics.*.retention` is a valid graphite retention specification

0.6.13 / 2017-08-21
    * Ensure `custom_metrics.*.aggregation` is a valid graphite aggregation specification

0.6.14 / 2017-08-21
    * Ensure `components.*.cluster` is either a boolean or a `{{repl ... }}` template

0.6.15 / 2017-08-21
    * Ensure `monitors.custom.*.display.label_scale` is a valid label scale expression

0.6.16 / 2017-08-21
    * Ensure `components.*.containers.*.content_trust.public_key_fingerprint` is a valid RFC4716 fingerprint
    * Ensure `images.*.content_trust.public_key_fingerprint` is a valid RFC4716 fingerprint

0.6.17 / 2017-08-21
    * Ensure `admin_commands.*.alias` is a valid shell alias
    * Ensure `properties.shell_alias` is a valid shell alias
    * Ensure `properties.logo_url` is a http or https url
    * Ensure a container's `volumes_from` section references existing containers
    * Ensure `name` entries are unique across all containers

0.6.18 / 2017-08-21
    * Schema fixes for swarm-specific options
    * Fix regex for config item type field

0.6.19 / 2017-08-21
    * Fix schema definition to allow boolean literals for
      `components.*.containers.*.volumes.*.is_ephemeral` and `.is_excluded_from_backup`

0.6.20 / 2017-08-29
    * Fix a bug that would allow URLs with spaces in the hostname

0.6.21 / 2017-08-29
    * Add new `test_proc` command `ldap_config_validate`

0.6.22 / 2017-09-05
    * Fix an issue where `when` clauses on config groups weren't being validated

0.6.23 / 2017-09-15
    * Tweak `mesg-yaml-empty` description, update docs generation.

0.6.24 / 2017-09-19
    * Ensure minimum Replicated API version of 2.8.0 when specifying a public port
    * Ensure swarm secrets have a name and value and that any labels have keys that are not the empty string

0.6.25 /2017-09-19
    * Schema fix for backup.hidden, now can be boolean or string. Formerly could only be string.

0.6.26 /2017-09-20
    * components.cluster_host_count and components.containers.cluster_instance_count uint properties are now checked to make sure they are actually uints
    * Schema fix for cluster_instance_count and cluster_host_count, all of their uint properties can now be strings or numbers

0.6.27 /2017-09-19
    * Check for swarm secrets label keys not "" now checks within an object, not an array

0.6.28 /2017-09-19
    * Schema fix so that ports[].public_port and ports[].private_port can be ints or strings, instead of only strings

0.6.29 /2017-09-19
    * Check that custom_requirements have unique ids

0.6.30 /2017-09-20
    * Validate `volumes_from` usage

0.6.31 /2017-09-21
    * Validate storage and memory sizes in host_requirements and kubernetes

0.6.32 /2017-09-25
    * Accept and validate a greater variety of admin command formats
    * Added a new predicate, `Not`

0.6.33 /2017-09-25
    * Verifies that if a public port is set, only one instance can run per cluster host

0.6.34 /2017-09-25

src/rules/root.ts
import { YAMLRule } from "../lint";

export const apiVersion: YAMLRule = {
  name: "prop-replicated-api-version-present",
  type: "error",
  message: "`replicated_api_version` must be present and be a valid Semver specification",
  test: { Semver: { path: "replicated_api_version", required: true } },
  examples: {
    wrong: [
      {
        description: "`replicated_api_version` is missing",
        yaml: `
---
{}
    `,
      },
      {
        description: "`replicated_api_version` is not valid semver",
        yaml: `
---
replicated_api_version: kfbr392
    `,
      },
    ],
    right: [{
      description: "`replicated_api_version` is valid semver",
      yaml: `
---
replicated_api_version: 2.9.0
      `,
    }],
  },
};

export const imageContentTrustValid: YAMLRule = {
  name: "prop-image-contenttrust-fingerprint-valid",
  type: "error",
  message: "An image's `content_trust.public_key_fingerprint` must be a valid RFC4716 fingerprint, e.g. `cb:69:19:cd:76:1f:17:54:92:a4:fc:a9:6f:a5:57:72`",
  test: {
    AnyOf: {
      path: "images",
      pred: {
        And: {
          preds: [
            { Truthy: { path: "content_trust.public_key_fingerprint" } },
            {
              NotMatch: {
                path: "content_trust.public_key_fingerprint",
                pattern: "[1-9a-f]{2}(:[1-9a-f]{2}){15}",
              },
            },
          ],
        },
      },
    },
  },
  examples: {
    wrong: [
      {
        description: "",
        yaml: `
---
images:
  - source: public
    name: redis
    tag: 3.2.1
    content_trust:
      public_key_fingerprint: flksdjflkds
    `,
      },
    ],
    right: [
      {
        description: "valid fingerprint",
        yaml: `
---
images:
  - source: public
    name: redis
    tag: 3.2.1
    content_trust:
      public_key_fingerprint: cb:69:19:cd:76:1f:17:54:92:a4:fc:a9:6f:a5:57:72
    `,
      },
    ],
  },
};

export const propertiesShellAlias: YAMLRule = {
  name: "prop-properties-shellalias-valid",
  type: "error",
  message: "`properties.shell_alias` must be a valid shell alias",
  test: {
    And: {
      preds: [
        { Truthy: { path: "properties.shell_alias" } },
        { NotMatch: { path: "properties.shell_alias", pattern: "^[a-zA-Z0-9_\\-]*$" } },
      ],
    },
  },
  examples: {
    wrong: [
      {
        description: "alias contains invalid character `&`",
        yaml: `
---
properties:
  shell_alias: exec&echo
      `,
      },
      {
        description: "admin command contains invalid character `*`",
        yaml: `
---
properties:
  shell_alias: exec**echo
      `,
      },
    ],
    right: [
      {
        description: "valid alias",
        yaml: `
---
properties:
  shell_alias: do_a-replicated_thing---plz
      `,
      },
    ],
  },
};

export const propertiesLogoURLValid: YAMLRule = {
  name: "prop-properties-logourl-valid",
  type: "error",
  message: "Logo URL must be a valid http or https URL",
  test: {
    And: {
      preds: [
        { Truthy: { path: "properties.logo_url" } },
        { InvalidURL: { path: "properties.logo_url" } },
      ],
    },
  },
  examples: {
    wrong: [
      {
        description: "protocol not in [`http`, `https`]",
        yaml: `
---
properties:
  logo_url: yo://mylogo.com/logo.png
      `,
      },
      {
        description: "invalid url",
        yaml: `
---
properties:
  logo_url: kfbr392
      `,
      },
      {
        description: "invalid url",
        yaml: `
---
properties:
  logo_url: http://i.goo gr.com/rnZ3Ftf.png
      `,
      },
    ],
    right: [
      {
        description: "valid url",
        yaml: `
---
properties:
  logo_url: http://x.y+a.com:3000/b/c
      `,
      },
    ],
  },
};

export const all: YAMLRule[] = [
  apiVersion,
  imageContentTrustValid,
  propertiesShellAlias,
  propertiesLogoURLValid,
];
